### PR TITLE
#2505 Allow to fork and disallow to create PRs for mirrors.

### PR DESCRIPTION
**Pull request by Download-Fritz**
Allow to fork and disallow to create PRs for mirrors, which addresses issue #2505.

Changes have been made to handle cases of prior created mirror repositories to provide backwards-compatibility.

Pull Requests were disabled, because...
  1) Mirrors are supposed to be read-only (merging a pull request would write to the it).
  2) prior to this change, no forks could have been made from within the GUI, making Pull Request support redundant.

### CODE REVIEW

File: models/repo.go [hunk 1]
```diff
@@ -931,6 +941,10 @@ func CreateRepository(u *User, opts CreateRepoOptions) (_ *Repository, err error
 		EnablePulls:  true,
 	}
 
+	if !repo.CanEnablePulls() {
+		repo.EnablePulls = false
```
Comment (Unknwon) [inline models/repo.go]: Isn't this redundant?

File: models/repo.go [hunk 2]
```diff
@@ -332,7 +332,12 @@ func (repo *Repository) IsOwnedBy(userID int64) bool {
 
 // CanBeForked returns true if repository meets the requirements of being forked.
 func (repo *Repository) CanBeForked() bool {
-	return !repo.IsBare && !repo.IsMirror
+	return !repo.IsBare
+}
+
+// CanEnablePulls returns true if repository meets the requirements of enabling pulls.
+func (repo *Repository) CanEnablePulls() bool {
```
Comment (Unknwon) [inline models/repo.go]: Maybe rename to `AllowPulls`?

**Comment by Download-Fritz** [comment]
@Unknwon Thank you for your feedback!

I left the CanEnablePulls() function in for the case the conditions need to be adapted in the future.
Though I implemented AllowsPulls() and got rid of the 'backwards-compatibility' code.

It indeed looks cleaner now. Thanks for yout feedback. ;)

File: routers/repo/issue.go [hunk 3]
```diff
@@ -48,7 +48,9 @@ func MustEnableIssues(ctx *middleware.Context) {
 }
 
 func MustEnablePulls(ctx *middleware.Context) {
-	if !ctx.Repo.Repository.EnablePulls {
+	if !ctx.Repo.Repository.CanEnablePulls() {
```
Comment (Unknwon) [inline routers/repo/issue.go]: Based on assumption that when `EnablePulls` is `true`, `CanEnablePulls` must be `true` as well, then this `if` check is redundant.

Comment (Download-Fritz) [inline routers/repo/issue.go]: @Unknwon That is not the case, though. Present mirror repositories will have EnablePulls set to true (as it is the default value), while mirrors should not support Pull Requests as of this commit.

File: routers/repo/issue.go [hunk 4]
```diff
@@ -48,7 +48,9 @@ func MustEnableIssues(ctx *middleware.Context) {
 }
 
 func MustEnablePulls(ctx *middleware.Context) {
```
Comment (Unknwon) [inline routers/repo/issue.go]: Then, this function should be renamed to `MustAllowPulls`.

File: routers/repo/issue.go (hunk 3 above)
Comment (Unknwon) [inline routers/repo/issue.go]: See my edited comments.

File: routers/repo/issue.go (hunk 4 above)
Comment (Download-Fritz) [inline routers/repo/issue.go]: @Unknwon You take clean code very seriously. :)
Fixed in a second commit. Thanks for the feedback.

**Comment by Unknwon** [comment]
Thanks!

**Pull request merged**